design/ip-address-management-for-networkdata.md

```diff
@@ -0,0 +1,328 @@
+      subnet: 192.168.0.0/24
+      step: 1
+    - start: 192.168.0.20
+      end: 192.168.0.25
+      subnet: 192.168.0.0/24
+status:
+  lastUpdated: "2020-04-02T06:36:09Z"
+  allocations:
+    "192.168.0.11": "metal3data-1"
+  addresses:
+    "metal3data-1": "pool-1-192-168-0-11"
+  error: false
+  errorMessage: ""
+```
+
+In the *spec* of *Metal3IPPool*, there would be a *pools* list, that would
+contain a list of IP address pools, with the *start* and *end* attributes
+giving the start and end ip addresses of the pool, the *subnet* allowing to
+verify that the allocated IP is in the pool and from which the start and end ip
+addresses can be inferred. The *step* item gives the interval between allocated
+IP addresses. Specifying single ip addresses can be achieved by setting the
+start and end ip address to that single ip address.
+
+The *status* would contain a *lastUpdated* field with the timestamp of the last
+update and and *error* boolean that is set to true if an error during allocation
+happens. In that case, the *errorMessage* will contain the text of the error.
+The *allocations* map will map the IP address to the *Metal3Data* object it was
+allocated for and the *addresses* will map the *metal3Data* objects with the
+*Metal3IPAddress* objects.
+
+All IP addresses specified in a Metal3IPPool must be from the same subnet.
+
+The *Metal3IPAddress* object would be the following
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3IPAddress
+metadata:
+  name: pool-1-192-168-0-11
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3IPPool
+    name: pool-1
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3Data
+    name: metal3data-1
+spec:
+  Metal3Data:
+    Name: metal3data-1
+  Address: 192.168.0.11
+  Metal3IPPool:
+    Name: pool-1
+status:
+  ready: true
+```
+
+For each owner reference added on a *Metal3IPPool* that is a *Metal3Data*, the
+controller reconciling the *Metal3IPPool* will select an available IP address.
+If all the *Metal3Data* owner references are from the same *Metal3DataTemplate*,
+then the controller will use the index of the *Metal3Data* to compute the IP
+address to allocate by sorting the addresses in incremental order and selecting
+the IP address corresponding to the index. In case of conflict, (multiple
+Metal3DataTemplate referencing the same *Metal3IPPool*, a random available IP
+address will be selected.
+
+Once the IP address is selected, the controller will create a *Metal3IPAddress*
+for that address. In case of conflict, it will list all the *Metal3IPAddress*
+objects that have an owner reference to this *Metal3IPPool* and update the
+status with the mapping of IP addresses and *Metal3Data* object names. It will
+then randomly select an available IP address. Once the *Metal3IPAddress* object
+is created, the *Metal3IPPool* object status will be updated with the new map.
```

Comment:
So it sound like the Metal3IPAddress CRs are the authoritative source of information about allocations, and the status fields are only updated as a convenience. Is that right? I wonder if we can do away with the status fields entirely?

Reply:
the status field is just used  as a cache to avoid fetching all Metal3IPAddress objects every time we are reconciling the Metal3IPPool. it's somehow an optimization, but we can do without it if you think it makes sense. It was implemented the same way for Metal3DataTemplate that has a cache in its status for Metal3Datas.